### ISSUE TITLE
Faster/Cheaper PBR Material

### ISSUE THREAD

**Issue by davehill00** [mrdoob/three.js#21578]
**Is your feature request related to a problem? Please describe.**

The PBR implementation in MeshStandardMaterial looks great, but is too expensive for devices like Oculus Quest/Quest 2. A scene that fills the entire screen with a quad using MeshStandardMaterial (with minimal features -- material color, a directional light, and an ambient light) renders at about 70Hz on Quest 2.

**Describe the solution you'd like**

I'd like to add a cheaper PBR implementation than MeshStandardMaterial -- perhaps [something like this](https://community.arm.com/cfs-file/__key/communityserver-blogs-components-weblogfiles/00-00-00-20-66/siggraph2015_2D00_mmg_2D00_renaldas_2D00_slides.pdf).

**Describe alternatives you've considered**

The alternative is using Phong material, but it's hard to translate PBR materials (e.g., from a GLTF file) into Phong materials.

**Additional context**

I'm planning to take a run at adding this kind of material to the ShaderLib/ShaderChunk system, but would certainly appreciate guidance or collaboration from folks who are more familiar with how this works. I'm still very new to the Three.JS codebase.

**Comment by liuyehua** [comment]
Good idea...

**Comment by Mugen87** [comment]
Related #14570.

**Comment by Ben-Mack** [comment]
Related: https://github.com/mrdoob/three.js/issues/19498

**Comment by arpu** [comment]
this is what hubs do https://github.com/mozilla/hubs/blob/master/src/utils/material-utils.js#L126 
works good to translate the MeshStandardMaterial from glb/gltf  to Phong Material

**Comment by elalish** [comment]
I'm definitely in favor of this. Not only could the Standard Material use some performance tuning, it could use a refactor/rewrite for code clarity. Still, this is likely to result in significant rendering differences, so I'd like to see this progress in parallel to the Standard Material and then when we get it to where it looks to be better, go ahead and replace the original. I work with the glTF standards group and I have [tools](https://modelviewer.dev/fidelity/) to help evaluate render fidelity and I'd be happy to give feedback. And @Ben-Mack's comment is also a good point: my PMREM could use some optimization too. Still, please don't resort to SH or dropping HDR; that leads to pretty terrible results for general lighting. If anyone has good ideas for how to perf benchmark these shaders or figure out where the bottlenecks are, that would be much appreciated. We don't want to over-index on a single GPU.

**Comment by donmccurdy** [comment]
> I'd like to see this progress in parallel to the Standard Material and then when we get it to where it looks to be better, go ahead and replace the original.

I'm not confident a material that is meaningfully cheaper to render (e.g. similar to Phong or Lambert) is ever going to look better than the existing MeshStandardMaterial. Do we have ideas for that kind of win/win optimization?

For example, a version of MeshStandardMaterial that does per-vertex lighting (like MeshLambertMaterial) will be much cheaper than the current PBR material, and easier to convert correctly to/from materials loaded from glTF. But certainly that will never replace the current MeshStandardMaterial.

A goal of "able to render the full screen with this material, at full device FPS" is probably a good goal, for some combination of VR and mobile devices.

**Comment by RemusMar** [comment]
> I'm not confident a material that is meaningfully cheaper to render (e.g. similar to Phong or Lambert) is ever going to look better than the existing MeshStandardMaterial. Do we have ideas for that kind of win/win optimization?

Maybe not better, but good enough:
https://necromanthus.com/Test/html5/Lara_mirror.html 
Click on the stage to switch between Phong and PBR.
**And good enough + faster is already a winner.**

**Comment by elalish** [comment]
I was looking at the first link from @davehill00, which is still physically-based (not Phong or Lambert), but with new approximations made for FP16 and such. I think those are well worth considering. Still, it's worth checking first to see if we're computation-bound or cache-bound. And it may well be there is more perf gain on the table from optimizing PMREM fetch than from all these new lighting approximations. Hence it'd be nice to have some numbers first.

**Comment by davehill00** [comment]
A bit more context from my end -- on the Oculus Browser team, we've been getting questions about rendering perf from teams using ThreeJS for VR experiences. I'd like to have an easy "out of the box" answer to point people at that will get them off on a reasonable path to maintaining acceptable performance, and the MeshStandardMaterial doesn't feel like the right path to recommend.

I had proposed "cheaper PBR" because most teams also seem to be relying on GLTF (which largely seems to assume PBR). Perhaps the suggestions around converting PBR material settings to Phong settings is good enough. If not, I wasn't proposing making changes to MeshStandardMaterial that degrade visual quality, but rather providing a lower-quality option that is still PBR but less demanding.

@elalish -- I'll try and get some data on Quest later this evening, but I'm pretty sure it's computation bound not cache-bound (I think you mean texture cache/texture bandwidth?) in my case (since I'm not using PMREM or any textures in my test scene).

**Comment by elalish** [comment]
> @elalish -- I'll try and get some data on Quest later this evening, but I'm pretty sure it's computation bound not cache-bound (I think you mean texture cache/texture bandwidth?) in my case (since I'm not using PMREM or any textures in my test scene).

For a realistic scene you'll probably want to include both IBL (PMREM) and material textures, at the very least baseColor, normalMap, and metallic-roughness. Without those you really aren't getting PBR at all and you could probably just convert to Phong as you say. But on the other hand, full PBR may not be needed for backgrounds and terrain, so a full screen quad may also be a bit extreme.

**Comment by davehill00** [comment]
@elalish Tried to do some investigation this evening, but my results are confusing so I'll need to do some more digging. 

The boxing experience I have here: https://davehill00.github.io/box/dist uses MeshStandardMaterial on floor, "red" wall, bag, gloves, etc. (but not the white walls). There are no lights in the scene, but it's using a PMREM to light things and also has baked lightmaps. This runs at a consistent 90Hz. I added a directional and ambient light, and it still runs at 90Hz.

I also have a simple room (just an inverted box with the starting point inside) and it runs consistently below 70Hz on the same headset. I've tried simpler and more complex materials, adding a PMREM, etc. and it's always slower than the boxing experience.

Can't explain what I'm seeing, so need to keep digging. :\

**Comment by mrdoob** [comment]
@davehill00 do you have a link to the 70Hz room?

**Comment by davehill00** [comment]
> @davehill00 do you have a link to the 70Hz room?

It's not up on GitHub yet. Will try to put something up tomorrow.

**Comment by davehill00** [comment]
@mrdoob here's the simple room scene that runs at ~68Hz on my Quest 2. Haven't had a chance to dig deeper into what's going on yet. 
https://davehill00.github.io/template/dist/

**Comment by davehill00** [comment]
Oof... okay, this is embarrassing. :( I had set a background image (cube map, from a 32-bit EXR) on the scene, which was rendering behind the wall but not contributing anything to the final scene. So that's where all the extra cycles were going, apparently. A little surprised it was that expensive, but definite explains the discrepancy between the simple scene and the more complex boxing one.

**Comment by FrostKiwi** [comment]
Sorry to necro an old thread with a blog post. Thought I add some fresh 2025 context with my Meta Quest 3 dev experience:

### Firmware was trash until 2024, all performance numbers need a new look
For some time, WebXR on Meta Quest 3 had horrendous performance issues stemming from a bad firmware update. It was the last big UI overhaul that happen, that kicked this off. It had to the way the rendering is done when in full VR / AR mode with no SystemUI overlay. It took over a year for this to be fixed, but it was fixed sometime in 2024. So all performance numbers taken before that aren't relevant anymore.

In general: In WebXR we don't get access to [Quest 3 Power-states](https://developers.meta.com/horizon/documentation/unreal/os-cpu-gpu-levels/) and are at the mercy of the `If possible, GPU level will increase when GPU utilization is 87% or greater, and decrease when GPU utilization is 81% or lesser.` rule. Also, we are locked to GPU power state 2 aka `456 MHz` in `immersive-ar`. Also no [`CPU Boost Hint`](https://developers.meta.com/horizon/documentation/native/android/po-quest-boost/#cpu-boost-hint) which I imagine would help the https://github.com/pmndrs/xr/issues/181 situation. The Quest 3 CPU can't even pull the per-frame Raycasts on 50k Poly meshes without dropping frames hard, that React Three Fiber XR sets up. Even in BVH supported non WebXR 2D Mode, CPU pegged at 100% util. So disabling Raycasts for the entire scene and using Collision meshes for all interactions is still the only way to get baseline performance.

### Shaders
Regarding the main topic of this Thread, Shaders:
Performance was also a problem for Meta when they did the [Project Flowerbed: A WebXR Case Study](https://developers.meta.com/horizon/blog/project-flowerbed-a-webxr-case-study/) in 2023.

> ### Shader Performance
> The best way to manage shader performance is to be aggressive about cutting features that aren’t significantly contributing to your scene’s visual quality. This means avoiding any wasted work! We made sure we weren’t paying any overhead costs for features we aren’t using. For example, the standard pixel shader for ***Three.js was calculating light probe contributions even though we didn’t have probes in the scene***.
We also turned off features we didn’t need based on each asset type—plants and the ground are set to maximum roughness and zero metallic, the ground under the water uses simplified lighting instead of the standard shader, and our sky uses the basic material. These kinds of optimizations are application specific—each developer should look at the visual contributions of the shaders they’re using and make a judgment call about what’s necessary.

Meta [seemed to have done a bunch of custom shader work](https://github.com/meta-quest/ProjectFlowerbed/tree/main/src/js/lib/shaders) with the "Three.js was calculating light probe contributions" seemingly refering to [this code passage](https://github.com/meta-quest/ProjectFlowerbed/blob/403a4f4d9748073893c6ff11313af1fd12371105/src/js/lib/shaders/WoodlandCSMShader.js#L232). Didn't cross-check with today's ThreeJS code, if that even matters today.

Downgrading to https://threejs.org/docs/#MeshPhongMaterial only would work in very stylized artstyles, so https://threejs.org/docs/#MeshStandardMaterial is still the priority for me personally.

As mentioned by @davehill00 
> A scene that fills the entire screen with a quad using MeshStandardMaterial

Yes, number one performance cap is the scene filling the field of view with `MeshStandardMaterial`. The GPU is hit hard.

Luckily, the Quest 3 hits 90 fps with `MeshStandardMaterial` Coverage of the whole field of view works for the Quest 3 with even complex geometry (500k poly scene) and moderate overdraw, but that is at standard Rendering resolution, which Meta capped *really* hard. The "Free" 4xMSAA enabled by default is a must. Forward Rendering is a must. Post-Processing, except the ThreeJS's built in (non post) ToneMapping is out of the Quest 3's performance budget.

[Virtual Desktop](https://www.vrdesktop.net/) (Still the best PC --> Quest 3 game Streaming tech) has some excellent numbers on this. They found that `3072x3216` per eye / `6144x3216` total gives an image the limit, at which 3D Scenes become indistinguishable "native" screen performance on the Quest 3, or rather the resolution that is needed to look the best after lens undistortion with not further improvement to be gained from higher resolutions. (What they call their VR Graphics Quality "God-Like (RTX 4090 / RX 9080-XT" Preset)

Meta set WebXR to render at `1680x1760` per eye `3360x1760` total (`frameBufferScaling` of 1.0,  foveation on by default), which is really not enough, giving off a "cheap VR feel" to everything. `frameBufferScaling` of 1.5 looks really good and matches the softness of `immersive-ar` passthrough, which lands us at `2520x2640` per eye `5040x2640` total. But with the GPU capped at performance Level 2, with MeshStandardMaterial, 90 fps with the scene covering the viewport is simply not possible. Async Warp will happen. Currently I default to `1.25` (`2100x2200` per eye, `4200x2200` total) at 90 fps for Browsers passing `/OculusBrowser/.test(navigator.userAgent);`

### `immersive-ar` and Async Warp don't mix
The 120 fps mode was not available at launch of the Quest 3 and came in a later firmware update. React Three Fiber XR https://pmndrs.github.io/xr/docs/tutorials/store defaults to this with their default `FrameRate` of `high`, instead of the 90fps the Quest 3 launched with, which is `medium` in React Three Fiber XR. Here React Three Fiber XR overules the Quest Brower's own 90fps default.

By my thinking, if a drop to asynchronous warp would to happen, it better happen to 60fps with 120fps base, not 45 fps with 90fps base. But ASync Warp and passthrough AR just don't mix well at all. If AsyncWarp happens, the desync between passthrough and rendered scene is horrendous. Instant nausea, looks bad on recordings due to desync.

Honestly, drop down to 72fps if you get any, because Async Warp messes with `immersive-ar` too much.

### Casting hits WebXR hard
One main thing is Recording and Casting to a screen, super important for client demos and stuff. But Casting to devices has a hard hit to WebXR performance, causing Async Warp, Causing desync with the Passthrough, causing nausea etc.. It also looks bad in recordings.
Dropping rendering Resolution to record videos and disabling Casting before giving a demo to the clients is what I do. The smooth experience matters most.

### My TL;DR:
- 1.25 Resolution scale
  - Meta's default is simply too low, lower than the already soft-ish passthrough, looks like your eyes can't focus.
- keep foveation, on by default it does its job
- Drop down to 72fps (passthrough ain't that faster anyways)
  - getting Async Warp is really that bad
  - No way the Quest 3 hits 120 fps (React Three Fiber XR default), 90fps is **usually** kept, but not always, with lots of `MeshStandardMaterial` filling the viewport
  - WebXR Experiences are not twitch shooters, Nausea Causing latency is less of a problem with passthrough acting as a nausea inhibitor for `immersive-ar`, so Resolution > Refreshrate.
- Result: 60% GPU Utilization at GPU Powerstate 2 with a 250k Scene filling the Quest 3 viewport with `MeshStandardMaterial` at 2x Overdraw

**Comment by mrdoob** [comment]
I just tried doing some changes: https://github.com/mrdoob/three.js/commit/3738cdb5807fe1613955c6e258cf64910fd00bc2

https://raw.githack.com/mrdoob/three.js/pbr-mobile/examples/index.html#webgl_animation_keyframes

But I don't see any improvements with the device I'm trying with (Mali-G72 MP3)

**Comment by FrostKiwi** [comment]
Fixing Quest 3 Position, `immersive-ar` @4200x2200 @72fps with Scene camera position to fill screen with couple of `MeshStandardMaterial` and switching out ThreeJS packages via adb + `chrome://inspect`.

Gotta clean up testing methodology make a nice Excel from the Oculus Debug .csv, but I think there *is* a measurable improvement, but tiny. So small, that I would need multiple runs to get an error spread to know whether that difference is within or outside measuring error.

Ignoring WebXR Entering, giving 10s for things to settle down, looking at a 10s window, ignoring, max and min GPU times:
| | 0.181.1 | https://github.com/mrdoob/three.js/commit/3738cdb5807fe1613955c6e258cf64910fd00bc2 |
| ----- | ----- | ------ |
| GPU Util % | Nailed to 67% | Nailed to 67% |
| App GPU Time in µs per frame | ~6630 µs | ~6560 µs |

<img width="960" height="540" alt="Image" src="https://github.com/user-attachments/assets/01865e88-b912-4e96-8ee9-817e3b057d70" />

**Comment by mrdoob** [comment]
Okay, here's a faster test: https://github.com/mrdoob/three.js/commit/fd0014763a64bdc5d82b6745e211f3978e63e4d0

https://raw.githack.com/mrdoob/three.js/pbr-mediump/examples/index.html#webgl_animation_keyframes

Lots of things may look broken but if performance doesn't improve then we know the shaders are not the issue.

**Comment by FrostKiwi** [comment]
@mrdoob pulled some numbers. Say if there are specific Metrics from the list that could nail down where to optimize.

<details><summary>Possible Metrics List</summary>

```
1       Clocks / Second
2       GPU Frequency
3       GPU % Bus Busy
4       Preemptions / second
5       Avg Preemption Delay
6       % Vertex Fetch Stall
7       % Texture Fetch Stall
8       % Texture L1 Miss
9       % Texture L2 Miss
10      % Stalled on System Memory
11      Read Total (Bytes/sec)
12      Write Total (Bytes/sec)
13      Texture Memory Read BW (Bytes/Second)
14      Vertex Memory Read (Bytes/Second)
15      SP Memory Read (Bytes/Second)
16      % Shaders Busy
17      % Shaders Stalled
18      % Shader ALU Capacity Utilized
19      % Time ALUs Working
20      % Time EFUs Working
21      % Texture Pipes Busy
22      % Wave Context Occupancy
23      % Instruction Cache Miss
24      % Time Shading Vertices
25      Vertices Shaded / Second
26      Vertex Instructions / Second
27      Textures / Vertex
28      Avg Bytes / Vertex
29      ALU / Vertex
30      EFU / Vertex
31      % Time Shading Fragments
32      Fragments Shaded / Second
33      Fragment Instructions / Second
34      Fragment ALU Instructions / Sec (Full)
35      Fragment ALU Instructions / Sec (Half)
36      Fragment EFU Instructions / Second
37      L1 Texture Cache Miss Per Pixel
38      Textures / Fragment
39      Avg Bytes / Fragment
40      ALU / Fragment
41      EFU / Fragment
42      % Time Compute
43      % Kernel Load Cycles
44      % Global Buffer Read L2 Hit
45      % Global Buffer Write L2 Hit
46      Global Memory Load Instructions
47      Global Memory Atomic Instructions
48      Global Memory Store Instructions
49      Global Buffer Data Read Request BW (Bytes/sec)
50      Global Buffer Data Read BW (Bytes/sec)
51      Global Image Data Read Request BW (Bytes/sec)
52      Global Image Uncompressed Data Read BW (Bytes/sec)
53      Global Image Compressed Data Read BW (Bytes/sec)
54      % L1 Hit
55      % Image Read L2 Hit
56      Local Memory Atomic Instructions
57      Local Memory Load Instructions
58      Local Memory Store Instructions
59      Load-Store Utilization
60      Avg Load-Store Instructions Per Cycle
61      Bytes Data Write Requested
62      Bytes Data Actually Written
63      Pre-clipped Polygons/Second
64      % Prims Trivially Rejected
65      % Prims Clipped
66      Average Vertices / Polygon
67      Reused Vertices / Second
68      Average Polygon Area
69      % Nearest Filtered
70      % Linear Filtered
71      % Anisotropic Filtered
72      % Non-Base Level Textures
73      % RTU Busy
74      % BVH Fetch Stall
75      Average BVH Fetch Latency Cycles
76      RTU Ray Box Intersections Per Instruction
77      RTU Ray Triangle Intersections Per Instruction
78      GPU % Utilization
79      % LRZ Busy
80      % LRZ Pixels Killed
81      % LRZ Prim Killed
```

</details>

Even though https://github.com/mrdoob/three.js/issues/32202 broke things visually due to `mediump`, https://github.com/mrdoob/three.js/commit/fd0014763a64bdc5d82b6745e211f3978e63e4d0 did not result in a visual change. I double checked my methodology by making sure https://github.com/mrdoob/three.js/tree/e1388ac56c6a13a97d4fd1268b75b80c227245e2 was weird and broken as it did in https://github.com/mrdoob/three.js/issues/32202, so it shouldn't be the case that I'm installing things wrong.

Today I learned, Meta Quests have great [low level debugging tools](https://developers.meta.com/horizon/documentation/unity/ts-ovrgpuprofiler/) with lot's of Shader related Metrics.
In the spirit of [Mitchell Hashimoto's "Can we test it? Yes, we can!"](https://youtu.be/MqC3tudPH6w?si=I3vLlbJS0mrOAmP2&t=358) and his automated GPU testing, with a sacrificial Meta Quest 3 one could totally built a Github Actions to gather these metrics 🤔

## 0.181.1
Accidentally exported Texture read Bandwith of Megabytes per Second as Billions per second 🤦‍♀️

<details><summary>Summary</summary>

| Metric | Minimum |Maximum | Average|
| ----------- | --- | --- | --- |
| GPU Percentage Utilization                             |           40.24  |           44.43   |          42.79 |
| L1 Texture Cache Miss Per Pixel                        |            0.24  |            0.25   |           0.25 |
| Fragments Shaded Per Second                            |  1.92 Billion/s  |  2.13 Billion/s   | 2.03 Billion/s |
| Fragment Instructions Per Second                       | 187.04 Billion/s |  205.44 Billion/s |  196.07 Billion/s |
| Fragment ALU Instructions Per Second (Full Precision)  | 147.48 Billion/s |  162.21 Billion/s |  154.98 Billion/s |
| Fragment ALU Instructions Per Second (Half Precision)  | 54.80 Billion/s  | 62.66 Billion/s  | 56.68 Billion/s |
| Fragment EFU Instructions Per Second                   | 12.13 Billion/s  | 13.31 Billion/s  | 12.75 Billion/s |
| Textures Per Fragment                                  |            1.80  |            1.90   |           1.85 |
| Percentage Time Shading Fragments                      |           92.67  |           93.24   |          93.07 |
| Texture Memory Read Bandwidth                          |  1.36 Billion/s  |  1.56 Billion/s   | 1.44 Billion/s |
| GPU Frequency                                          |         456 MHz  |         456 MHz   |        456 MHz |
| Percentage Shaders Busy                                |           64.86  |           67.80   |          66.27 |

</details>

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/b9f63807-86f1-49f9-91fe-d1a5d92f7926" />

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/f51c2434-5b61-4c0b-88d6-2c97d1ddee2d" />

## https://github.com/mrdoob/three.js/commit/3738cdb5807fe1613955c6e258cf64910fd00bc2

<details><summary>Summary</summary>

| Metric | Minimum |Maximum | Average|
| ----------- | --- | --- | --- |
| GPU Percentage Utilization                            |            41.08  |           44.31   |          42.73 |
| L1 Texture Cache Miss Per Pixel                       |             0.24  |            0.25   |           0.25 |
| Fragments Shaded Per Second                           |   1.96 Billion/s  |  2.11 Billion/s   | 2.03 Billion/s |
| Fragment Instructions Per Second                      |  191.63 Billion/s |  210.47 Billion/s |  198.44 Billion/s |
| Fragment ALU Instructions Per Second (Full Precision) |  151.56 Billion/s |  167.45 Billion/s |  157.02 Billion/s |
| Fragment ALU Instructions Per Second (Half Precision) |  54.81 Billion/s  | 61.90 Billion/s  | 56.72 Billion/s |
| Fragment EFU Instructions Per Second                  |  12.61 Billion/s  | 13.86 Billion/s  | 13.06 Billion/s |
| Textures Per Fragment                                 |             1.81  |            1.89   |           1.85 |
| Percentage Time Shading Fragments                     |            92.67  |           93.20   |          93.00 |
| Texture Memory Read Bandwidth                         |   1.36 Billion/s  |  1.55 Billion/s   | 1.43 Billion/s |
| GPU Frequency                                         |          456 MHz  |         456 MHz   |        456 MHz |
| Percentage Shaders Busy                               |            64.62  |           67.36   |          65.81 |

</details>

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/8b5299b2-b043-41c7-b175-70997aaa8fcf" />

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/50846e5e-ccd2-4ecc-8a3f-d1204b84ff9f" />

## https://github.com/mrdoob/three.js/commit/fd0014763a64bdc5d82b6745e211f3978e63e4d0

<details><summary>Summary</summary>

| Metric | Minimum |Maximum | Average|
| ----------- | --- | --- | --- |
| GPU Percentage Utilization                             |           40.30  |           44.11   |          42.37 |
| L1 Texture Cache Miss Per Pixel                        |            0.24  |            0.25   |           0.25 |
| Fragments Shaded Per Second                            |  1.97 Billion/s  |  2.13 Billion/s   | 2.03 Billion/s |
| Fragment Instructions Per Second                       | 192.14 Billion/s |  208.20 Billion/s |  198.51 Billion/s |
| Fragment ALU Instructions Per Second (Full Precision)  | 152.07 Billion/s |  165.15 Billion/s |  157.10 Billion/s |
| Fragment ALU Instructions Per Second (Half Precision)  | 54.73 Billion/s  | 62.58 Billion/s  | 56.69 Billion/s |
| Fragment EFU Instructions Per Second                   | 12.65 Billion/s  | 13.68 Billion/s  | 13.07 Billion/s |
| Textures Per Fragment                                  |            1.81  |            1.89   |           1.85 |
| Percentage Time Shading Fragments                      |           92.69  |           93.45   |          93.05 |
| Texture Memory Read Bandwidth                          |  1.37 Billion/s  |  1.54 Billion/s   | 1.43 Billion/s |
| GPU Frequency                                          |         456 MHz  |         456 MHz   |        456 MHz |
| Percentage Shaders Busy                                |           65.31  |           68.55   |          66.64 |

</details>

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/b16fcb30-53cf-4993-9839-82a5439a41c0" />

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/6da1ce65-616d-49cb-8b79-9c0e4df4f749" />

## `PMREM 2048` https://github.com/mrdoob/three.js/commit/e1388ac56c6a13a97d4fd1268b75b80c227245e2

<details><summary>Summary</summary>

| Metric | Minimum |Maximum | Average|
| ----------- | --- | --- | --- |
| GPU Percentage Utilization                             |           41.27  |           44.91    |          42.91 |
| L1 Texture Cache Miss Per Pixel                        |            0.24  |            0.25    |           0.25 |
| Fragments Shaded Per Second                            |  1.97 Billion/s  |  2.13 Billion/s    | 2.03 Billion/s |
| Fragment Instructions Per Second                       | 189.74 Billion/s |  203.89 Billion/s  |  196.44 Billion/s |
| Fragment ALU Instructions Per Second (Full Precision)  | 150.00 Billion/s |  163.14 Billion/s  |  155.31 Billion/s |
| Fragment ALU Instructions Per Second (Half Precision)  | 54.81 Billion/s  | 62.60 Billion/s   | 56.71 Billion/s |
| Fragment EFU Instructions Per Second                   | 12.33 Billion/s  | 13.31 Billion/s   | 12.77 Billion/s |
| Textures Per Fragment                                  |            1.79  |            1.91    |           1.85 |
| Percentage Time Shading Fragments                      |           92.78  |           93.28    |          93.09 |
| Texture Memory Read Bandwidth                          |  1.39 Billion/s  |  1.61 Billion/s    | 1.46 Billion/s |
| GPU Frequency                                          |         456 MHz  |         456 MHz    |        456 MHz |
| Percentage Shaders Busy                                |           65.05  |           67.41    |          66.32 |

</details>

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/26921f3c-adf3-4cc9-90b5-6abb5a484851" />

<img width="48%" alt="Image" src="https://github.com/user-attachments/assets/e165ea8c-e52f-4147-9cfe-6ad35c473c4b" />

**Comment by mrdoob** [comment]
Looks like this will require more fine grained tests...

I'm thinking of getting a Pixel 2XL, which should be equivalent to a Quest 1, and see if I can optimize things on it.

**Comment by Mugen87** [comment]
I did a performance analysis lately on our PBR material and it seems optimizing PMREM is one of the potential biggest levers. I'm not super into the details of PMREM but would it be possible to move away from the current atlas approach and represent the PMREM as a real (mip-mapped) cube map? 

Ideally, we use just a single instruction `textureLod( envMap, envMapRotation * reflectVec, lod )` to sample the PMREM. That would replace most of what happens inside `textureCubeUV()`, meaning face selection, uv projection, atlas addressing, seam handling, manual trilinear filtering. AFAICS, all this only exists to emulate what the texture unit does natively for a cubemap.

I'm not sure but does the current atlas approach come form a time where WebGL1 compatibility was required? Now with WebGL 2 and WebGPU, we can safely rely on `textureLod()` and `textureSampleLevel()`.

**Comment by mrdoob** [comment]
Feel free to give it a try yeah 👍
